rust-keybroker/keybroker-client/src/lib.rs

```diff
@@ -1,17 +1,458 @@
+    0xeb, 0xe6, 0xe7, 0xd9, 0x65, 0x4b, 0xbe, 0x86, 0x4e, 0x30, 0x0d, 0x49, 0x19, 0xac, 0xcf, 0x84,
+    0x58, 0x20, 0x24, 0xd5, 0xb0, 0xa2, 0x96, 0xcc, 0x05, 0xcb, 0xd8, 0x06, 0x8c, 0x50, 0x67, 0xc5,
+    0xbd, 0x47, 0x3b, 0x77, 0x0d, 0xda, 0x6a, 0xe0, 0x82, 0xfe, 0x3b, 0xa3, 0x0a, 0xbe, 0x3f, 0x9a,
+    0x6a, 0xb1, 0x58, 0x20, 0x78, 0x8f, 0xc0, 0x90, 0xbf, 0xc6, 0xb8, 0xed, 0x90, 0x31, 0x52, 0xba,
+    0x84, 0x14, 0xe7, 0x3d, 0xaf, 0x5b, 0x8c, 0x7b, 0xb1, 0xe7, 0x9a, 0xd5, 0x02, 0xab, 0x06, 0x99,
+    0xb6, 0x59, 0xed, 0x16, 0x58, 0x20, 0xda, 0xc4, 0x6a, 0x58, 0x41, 0x5d, 0xc3, 0xa0, 0x0d, 0x7a,
+    0x74, 0x18, 0x52, 0x00, 0x8e, 0x9c, 0xae, 0x64, 0xf5, 0x2d, 0x03, 0xb9, 0xf7, 0x6d, 0x76, 0xf4,
+    0xb3, 0x64, 0x4f, 0xef, 0xc4, 0x16, 0x58, 0x20, 0x32, 0xc6, 0xaf, 0xc6, 0x27, 0xe5, 0x55, 0x85,
+    0xc0, 0x31, 0x55, 0x35, 0x9f, 0x33, 0x1a, 0x0e, 0x22, 0x5f, 0x68, 0x40, 0xdb, 0x94, 0x7d, 0xd9,
+    0x6e, 0xfa, 0xb8, 0x1b, 0xe2, 0x67, 0x19, 0x39, 0x58, 0x60, 0x6e, 0xaf, 0x8c, 0x12, 0xcd, 0x0f,
+    0xc0, 0xd6, 0x2b, 0x3d, 0x3b, 0x84, 0xa1, 0x6a, 0x6b, 0x76, 0x28, 0xdb, 0xc8, 0xe3, 0x46, 0xfb,
+    0x9f, 0x09, 0x44, 0xda, 0x5a, 0x81, 0xd0, 0x09, 0x32, 0x33, 0x9c, 0x6d, 0x6b, 0xd1, 0xf8, 0x91,
+    0xea, 0x64, 0xc0, 0xb0, 0x42, 0xf5, 0x1c, 0xa3, 0x31, 0x87, 0xed, 0xbb, 0xe7, 0xf0, 0xa3, 0x95,
+    0x98, 0x55, 0x06, 0x89, 0x09, 0x3b, 0x0d, 0x99, 0xbb, 0xf3, 0x8e, 0x93, 0x2c, 0x5e, 0x2e, 0x18,
+    0x26, 0xc4, 0xdb, 0xc3, 0xed, 0x3e, 0x49, 0x44, 0x18, 0xb0, 0x4f, 0x79, 0x1d, 0x88, 0xda, 0x64,
+    0xec, 0x3a, 0x3f, 0x16, 0x50, 0x96, 0x51, 0xe7, 0x30, 0x97,
+];
+
+/// An EvidenceProvider mock.
+///
+/// The CcaExampleToken implementation of the EvidenceProvider trait is really
+/// nothing but a mock, ignoring the challenge and returning the static evidence
+/// that is used in the CCA exemple. This is useful during bring-up or debug to
+/// un-tie the link between the key request and the evidence submission.
+pub struct CcaExampleToken {}
+
+impl EvidenceProvider for CcaExampleToken {
+    fn get_evidence(&self, _challenge: &str, _verbose: bool) -> Result<Vec<u8>> {
+        Ok(CCA_EXAMPLE_TOKEN.to_vec())
+    }
 }
 
-#[cfg(test)]
-mod tests {
-    use super::*;
+/// A TSM attestation report implementation of EvidenceProvider.
+///
+/// The TsmAttestationReport implementation of the EvidenceProvider trait uses
+/// Linux's TSM attestation report infrastructure to construct an evidence from
+/// a challenge.
+pub struct TsmAttestationReport {}
+
+impl EvidenceProvider for TsmAttestationReport {
+    fn get_evidence(&self, challenge: &str, verbose: bool) -> Result<Vec<u8>> {
+        match TsmReportPath::new(TsmReportProvider::Cca) {
+            Ok(tsm_report_path) => match URL_SAFE_NO_PAD.decode(challenge) {
+                Ok(challenge) => {
+                    if verbose {
+                        println!("Challenge ({} bytes) = {:02x?}", challenge.len(), challenge);
+                    }
+                    assert_eq!(
```

Comment:
Library code should only assert internal invariants.
An invalid input parameter should trigger an error.

Reply:
Isn't the challenge length of 64 byte an (internal) invariant ? But yeah, I will update the patch to not assert and report some meaningfull error instead.

Reply:
Patch updated !

Reply:
> Isn't the challenge length of 64 byte an (internal) invariant ?

IIRC, the RSI ABI states that the input can be _at most_ 64B, and anything less than that will be zero-padded to 64.

> But yeah, I will update the patch to not assert and report some meaningfull error instead.

Thanks!